File: ruby/graphs/breadth_first_search.rb
require_relative './graph.rb'

module Graphs
  class BreadthFirstSearch
    class << self

      # Internal: Traverses through the vertices of a graph in a level wise,
      #           one level at a time manner
      #           Next level is considered only after all the elements in the
      #           current level are traversed through
      #
      # graph - Graph to be searched
      # s - Starting node in the graph from which the search has to be initiated
      # searchable - Node in the graph to be searched for
      #
      # Examples
      #   BFS(graph, graph.vertices.first, graph.vertices.last)
      #
      def BFS(graph, s, searchable = nil)
        graph.vertices.select { |x| x != s }.each do |u|
          u.color = 'WHITE'
          u.d = Float::INFINITY
          u.pi = nil
        end

        s.color = 'GRAY'
        s.d = 0
        s.pi = nil

        q = Queue.new
        q.enq(s)
        while q.size != 0
          u = q.deq
          u.adj_list.each do |v|
            if v.color == 'WHITE'
              return "FOUND!!" if v.attribute == searchable.attribute
              v.color = 'GRAY'
              v.d = u.d + 1
              v.pi = u
              q.enq(v)
            end
          end
          u.color = 'BLACK'
        end
      end

      def print_BFS_path(g, s, v)
        if v == s
          print s.attribute
        elsif v.pi == nil
          print "NO PATH FROM #{s.attribute} TO #{v.attribute} EXISTS"
        else
          print_BFS_path(g, s, v.pi)
          print v.attribute
        end
      end

    end
  end
end

File: Java/GraphAlgorithms/BreadthFirstSearch.java
package graphalgorithms;

import java.util.List;
import java.util.LinkedList;
import java.util.Queue;

public class BreadthFirstSearch {

    public void bfs(Graph g, Vertex s) {
        if (g == null || s == null) {
            throw new IllegalArgumentException("Graph object or Vertex object null");
        }
        List<Vertex> vertices = g.getVertices();
        for (Vertex u : vertices) {
            u.color = Color.WHITE;
            u.d = -1;
            u.p = null;
        }
        s.color = Color.GRAY;
        s.d = 0;
        s.p = null;

        Queue<Vertex> q = new LinkedList<>();
        q.add(s);
        System.out.printf("BFS: ");
        while (!q.isEmpty()) {
            Vertex u = q.poll();
            System.out.printf("%s ", u);
            List<Vertex> adjVertices = u.neighbors;
            for (Vertex v : adjVertices) {
                if (v.color == Color.WHITE) {
                    v.color = Color.GRAY;
                    v.d = u.d + 1;
                    v.p = u;
                    q.add(v);
                }
            }
            u.color = Color.BLACK;
        }
        System.out.printf("%n");
    }

    public void printPath(Graph g, Vertex s, Vertex v) {
        if (g == null || s == null || v == null) {
            throw new IllegalArgumentException("Graph object or Vertex object(s) null");
        }
        if (v == s) {
            System.out.printf("%s%n", s);
        } else if (v.p == null) {
            System.out.printf("no path from %s to %s exists.%n", s, v);
        } else {
            printPath(g, s, v.p);
            System.out.printf("%s%n", v);
        }
	
    }

}